jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/JavaxWebSocketBasicRemote.java

```diff
@@ -65,10 +65,10 @@ public void sendBinary(ByteBuffer data) throws IOException
         {
             LOG.debug("sendBinary({})", BufferUtil.toDetailString(data));
         }
-        try (SharedBlockingCallback.Blocker b = session.getBlocking().acquire())
-        {
-            sendFrame(new Frame(OpCode.BINARY).setPayload(data), b, false);
-        }
+
```

Comment:
The use of the SharedBlockingCallback provided both mutual exclusion (waiting for a previously allocated blocker) as well as the BlockingCallback itself.

It turns out that the mutual exclusion is vulnerable to reentrancy, so that if the blocking operation calls another operation that needs the same SharedBlockingCallback, then we get deadlock.

We have seen that close can have that reentrant behaviour, as it calls user code, but I'm not sure that means we should give up on the mutual exclusion elsewhere.  For example are we sure we can allow two threads to simultaneously enter `sendBinary`?

Thus I'm not sure we should swap out all the SharedBlockingCallbacks, specially as we don't really know where else this reentrancy might be a problem.  So I think we need to at least modify SharedBlockingCallback to detect such reentrancy and to either do an ISE or perhaps optionally provide a non shared instance.   Perhaps in the methods we know that might be reentrant, we can pass in an extra arg to say that renentrancy is supported:
```
        try (SharedBlockingCallback.Blocker b = session.getBlocking().acquire(true))	
```
In this case, if the Blocker is already allocated, then we simply return a new instance that is never shared.

Reply:
@gregw good point! But perhaps `AutoLock` is enough then?

```java
void send(ByteBuffer) 
{
  try (AutoLock lock = _lock.lock()) // Provides exclusion.
  {
    BlockingCallback c = new BlockingCallback();
    send(..., c);
    c.block(); // Provides blocking semantic.
  }
}
```

Concurrent close from different threads is now handled by the `AtomicBoolean` guard so it won't deadlock.

Reply:
Adding locks would return the mutual exclusion, but I'm kind of concerned that we've never noticed that `SharedBlockingCallback` is not reentrant before.
So I'd rather fix `SharedBlockingCallback` to a) detect and warn of such reentrancy; b) to handle it by providing a non-shared version.  That way we would not need to change code in many places.

Reply:
Why do we need this mutual exclusion for things like `sendBinary` I can't see why there is any advantage to this over letting core handle it.

With only using the `BlockingCallback` the entry gets queued in the `WebSocketCoreSession` flusher and then the thread blocks on the Callback until the frame is sent.

With the `SharedBlockingCallback` the thread blocks trying to get the `Callback` itself, then when it gets the callback it can give it to the CoreSession it goes in the flusher and we block on the Callback again.

It wouldn't take more memory to just put the frame in the flusher queue.

Reply:
Chatted to @sbordet  about this.
We kind of arrived at the conclusion that the websocket code doesn't need mutual exclusion, so your approach of avoiding using `SharedBlockingCallback` is good... especially as jetty-9 doesn't mutually exclude and just throws a pending exception. We should probably review it's usage everywhere... but that's another issue!

However, we probably do need your BlockingCallback to have an idleTimeout mechanism to avoid waiting forever if the callback fails